onnxruntime/core/optimizer/transpose_optimizer/layout_tranformer_dev_notes.md

```diff
@@ -0,0 +1,53 @@
+# Layout Transformer Dev Notes
+
+## Overview
+ONNX standard assumes NCHW format for tensor layout. However, NCHW is not the best (perf efficient) format for all hardware types. Depending on the underlying hardware, to get the best perf we need to convert the model or in some cases part of the model which will be executed on the hardware from NCHW -> NHWC. Layout Transformer enables just this. It works with ONNX and ORT format models.
+
+*Note: Currently Layout Transformer works only for compiling EPs like NNAPI, QNN EP etc... More work is needed for it to be compatible with EPs like CPU and CUDA which use static kernel registration.*
+
+## How it works
+Layout transformer is [invoked during graph partitioning](https://github.com/microsoft/onnxruntime/blob/1a4868e5c4c4a270ad91036e36f2a03410c4c278/onnxruntime/core/framework/graph_partitioner.cc#L111) phase. After getting capabilities from the registered EP, layout transformer converts the part of the graph which is claimed by the EP to NHWC IF this is the desired format for the given EP.
+
+```
+Foreach
+  1. call GetCapability
+  2. IF EP.DesiredFormat == NHWC
+    2.1. Invoke Layout Transformer
+    2.2 If graph is modified -> call GetCapability (layout transformer can add new nodes to the graph)
+  3 Compile
+```
+
+Layout Transformer does multiple top to bottom passes on the graph in order to produce the most efficient graph with as little transpose ops as possible.
+
+*IMPORTANT NOTE* After layout transformation is done, graph resolve cannot be called for the graph. This is because graph resolve validates the shape of the nodes by calling ONNX TypeAndShapeInferenceFunction, these type and shape inf functions can ONLY infer shapes for NCHW format values. Therefore, when passed a graph with NHWC nodes the inferred shape validation fails and hence graph resolve throws. This is the very reason layout transformation is *NOT ENABLED* when Graph Partitioner Mode is kAssignOnly.
+
+### Convert layout for applicable nodes
+[This is the first pass](https://github.com/microsoft/onnxruntime/blob/1a4868e5c4c4a270ad91036e36f2a03410c4c278/onnxruntime/core/optimizer/transpose_optimizer/optimizer_api_impl.cc#L815). Layout transformer simply inserts transpose nodes before and after every layout sensitive ops that are claimed by the EP. After this pass, the graph is correct but extremely inefficient.
```

Comment:
Might help to be more explicit and say it inserts an NCHW->NHWC Transpose before the node and an NHWC->NCHW Transpose after the node.